### PR TITLE
feat: implement `transfer_from` and `decrease_allowance`

### PR DESCRIPTION
**Pull request by chungquantin**
Parent PR: https://github.com/r0gue-io/pop-node/pull/113
ISSUE: https://github.com/r0gue-io/pop-node/issues/104

Original wasm size: 30.7K, Optimized: 7.2K

## Tasks
`transfer_from`
- [x] `pallet-api` implementation
- [x] pallet unit testing
- [x] contract integration test 

`decrease_allowance`
- [x] `pallet-api` implementation
- [x] pallet unit testing
- [x] contract integration test 
- [ ] benchmarking

### CODE REVIEW

**Review by chungquantin — COMMENTED** [review]
Verify non-impacted code due to reformatting

**Comment by Daanvdplas** [comment]
Could you please use the old formatting `cargo fmt +stable` for now? It creates a lot of noise in the PR and reviewers are already struggling with the size of the PR. We will change the formatting as soon as the api is merged in. As for your formatting for comments, that can stay of it is not automated.

**Comment by chungquantin** [comment]
@Daanvdplas Our current `rustfmt.toml` config params are not supported by the stable channel at the moment

```
➜ cargo +stable fmt --check 

Warning: can't set `wrap_comments = true`, unstable features are only available in nightly channel.
Warning: can't set `comment_width = 100`, unstable features are only available in nightly channel.
Warning: can't set `imports_granularity = Crate`, unstable features are only available in nightly channel.
Warning: can't set `reorder_impl_items = false`, unstable features are only available in nightly channel.
Warning: can't set `spaces_around_ranges = false`, unstable features are only available in nightly channel.
Warning: can't set `binop_separator = Back`, unstable features are only available in nightly channel.
Warning: can't set `match_arm_blocks = false`, unstable features are only available in nightly channel.
Warning: can't set `trailing_semicolon = false`, unstable features are only available in nightly channel.
Warning: can't set `trailing_comma = Vertical`, unstable features are only available in nightly channel.
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 1]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		);
+		assert_eq!(decoded::<Error>(unapproved_result), Module { index: 52, error: 10 },);
+
+		// Check if the allowance is correct
+		assert_eq!(0, Assets::allowance(asset, &owner, &delegate.clone()));
+		assert!(
+			!approve(owner.clone(), asset, delegate.clone(), amount).did_revert(),
+			"Contract reverted!"
+		);
+		assert_eq!(Assets::allowance(asset, &owner, &delegate.clone()), amount);
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:450]: Only call the contract message that we are testing. Other calls should be made via the pallet directly. In addition, allowance checks are unnecessary.

Alice approves the contract to transfer on behalf of it:
```suggestion
	assert_ok!(Assets::approve_transfer(
		RuntimeOrigin::signed(ALICE.into()),
		asset_id.into(),
		addr.clone().into(),
		amount,
	));
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 2]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:451]:
```suggestion
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 3]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		// Successfully transfer
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:452]:
```suggestion
		// Successful transfer.
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 4]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		);
+
+		// Allow `delegate` to spend `amount` owned by contract address
+		let asset = create_asset_and_mint_to(ALICE, ASSET_ID, owner.clone(), amount * 2);
+		// `delegate` transfer from the `owner` without approval
+		let unapproved_result = transfer_from(
+			addr.clone(),
+			asset,
+			delegate.clone(),
+			owner.clone(),
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:437]:
```suggestion
			ALICE,
```

File: pallets/api/src/mock.rs [hunk 5]
```diff
@@ -102,6 +102,8 @@ impl crate::fungibles::Config for Test {
 
 pub(crate) const ALICE: AccountId = 1;
 pub(crate) const BOB: AccountId = 2;
+pub(crate) const CHARLIE: AccountId = 3;
+pub(crate) const DAVE: AccountId = 4;
```
Comment (Daanvdplas) [inline pallets/api/src/mock.rs:106]: We don't need them, see suggestions in `transfer_from_works()`

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 6]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		let owner = addr.clone();
+		let delegate = CHARLIE;
+		let asset = ASSET_ID;
+
+		// Asset does not exist
+		assert_eq!(
+			decoded::<Error>(transfer_from(
+				addr.clone(),
+				asset,
+				delegate.clone(),
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:421]:
```suggestion
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 7]
```diff
@@ -79,6 +85,23 @@ fn transfer(
 	result
 }
 
+fn transfer_from(
+	addr: AccountId32,
+	asset_id: AssetId,
+	delegate: AccountId32,
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:91]: remove

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 8]
```diff
@@ -79,6 +85,23 @@ fn transfer(
...
+	from: AccountId32,
+	to: AccountId32,
+	value: Balance,
+	data: Vec<u8>,
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:95]: remove

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 9]
```diff
@@ -79,6 +85,23 @@ fn transfer(
...
+) -> ExecReturnValue {
+	let function = function_selector("transfer_from");
+	let params =
+		[function, asset_id.encode(), from.encode(), to.encode(), value.encode(), data.encode()]
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:99]: For simplicity:

```suggestion
		[function, asset_id.encode(), from.encode(), to.encode(), value.encode(), vec![u8]]
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 10]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		let bob_balance_before_transfer = Assets::balance(asset, &BOB);
+		let approved_result = transfer_from(
+			addr.clone(),
+			asset,
+			delegate.clone(),
+			owner.clone(),
+			BOB,
+			amount / 2,
+			vec![],
+		);
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:462]: Contract has been approved by Alice. The contract will now transfer `asset` from `ALICE` to `BOB`

```suggestion
		let approved_result = transfer_from(addr.clone(), asset, ALICE, BOB, amount / 2);
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 11]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+				delegate.clone(),
+				ALICE,
+				BOB,
+				1,
+				vec![]
+			)),
+			Module { index: 52, error: 3 },
+		);
+
+		// Allow `delegate` to spend `amount` owned by contract address
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:430]: This is old I think

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 12]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+		let asset = create_asset_and_mint_to(ALICE, ASSET_ID, owner.clone(), amount * 2);
+		// `delegate` transfer from the `owner` without approval
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:432]:
```suggestion
		// Unapproved
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 13]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
+				ALICE,
+				BOB,
+				1,
+				vec![]
+			)),
+			Module { index: 52, error: 3 },
+		);
+
+		// Allow `delegate` to spend `amount` owned by contract address
+		let asset = create_asset_and_mint_to(ALICE, ASSET_ID, owner.clone(), amount * 2);
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:431]:
```suggestion
		let asset = create_asset_and_mint_to(ALICE, ASSET_ID, ALICE, amount * 2);
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 14]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
...
 
+#[test]
+fn transfer_from_works() {
+	new_test_ext().execute_with(|| {
+		let _ = env_logger::try_init();
+		let addr = instantiate("contracts/fungibles/target/ink/fungibles.wasm", INIT_VALUE, vec![]);
+		let amount: Balance = 100 * UNIT;
+		let owner = addr.clone();
+		let delegate = CHARLIE;
+		let asset = ASSET_ID;
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:414]:
```suggestion
```

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 15]
```diff
@@ -367,6 +403,69 @@ fn transfer_works() {
 	});
 }
 
+#[test]
+fn transfer_from_works() {
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/local_fungibles.rs:407]: Could we add test situations for `AssetNotLive`. See other tests how to (freeze and destroy)

File: pop-api/integration-tests/src/local_fungibles.rs [hunk 16]
```diff
@@ -14,6 +15,11 @@ fn decoded<T: Decode>(result: ExecReturnValue) -> T {
 	}
 }
 
+fn checked_decoded<T: Decode>(result: ExecReturnValue) -> Result<T, String> {
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pop-api/integration-tests/src/local_fungibles.rs:18]: After this PR is merged, I will create another PR to refractor.

[the remark this answers is not in the document]
Comment (Daanvdplas) [reply pop-api/integration-tests/src/local_fungibles.rs:18]: Please remove and it to #120

**Comment by Daanvdplas** [comment]
This should be waited for review until #113 is approved

**Pull request closed without merging**